SI-8177 co-evolve more than just RefinedTypes

adriaanm opened the pull request:
`asSeenFrom` produced a typeref of the shape `T'#A` where `A` referred to a symbol
defined in a `T` of times past.

More precisely, the `TypeRef` case of `TypeMap`'s `mapOver` correctly modified a prefix
from having an underlying type of `{ type A = AIn }` to `{ type A = Int }`,
with a new symbol for `A` (now with info `Int`), but the symbol in the outer
`TypeRef` wasn't co-evolved (so it still referred to the `A` in `{ type A = AIn }`
underlying the old prefix).

`coEvolveSym` used to only look at prefixes that were directly `RefinedType`s,
but this bug shows they could also be `SingleType`s with an underlying `RefinedType`.

Review by @paulp

Ichoran committed:
SI-6736 Range.contains is wrong

Removed once-used private method that was calculating ranges in error and corrected the contains method (plus improved performance).

Ichoran committed:
Reasonable Range operations consistently work when overfull.

Operations are reasonable when they don't require indexing or conversion into a collection.  These include head, tail, init, last, drop, take, dropWhile, takeWhile, dropRight, takeRight, span.

Tests added also to verify the new behavior.

adriaanm committed:
SI-8177 co-evolve more than just RefinedTypes

`asSeenFrom` produced a typeref of the shape `T'#A` where `A` referred to a symbol
defined in a `T` of times past.

More precisely, the `TypeRef` case of `TypeMap`'s `mapOver` correctly modified a prefix
from having an underlying type of `{ type A = AIn }` to `{ type A = Int }`,
with a new symbol for `A` (now with info `Int`), but the symbol in the outer
`TypeRef` wasn't co-evolved (so it still referred to the `A` in `{ type A = AIn }`
underlying the old prefix).

`coEvolveSym` used to only look at prefixes that were directly `RefinedType`s,
but this bug shows they could also be `SingleType`s with an underlying `RefinedType`.

paulp commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -2188,13 +2188,16 @@ trait Types
+    // However, there are other ways a typeref's symbol can refer to a symbol defined in its prefix
+    // for example, if `preNew` is a `SingleType`s with an underlying `RefinedType`.
+    // To deal with erroneous `preNew`, fallback via `orElse sym` in case `preNew` does not have a decl named `sym.name`
+    // (I don't think it's possible for the declaration to be inherited, so we can use `decls.lookup` instead of `member`.)
```
Why isn't it possible for it to be inherited?

Can we perhaps agree that some logging code is in order here? Test your assumptions?

adriaanm replied:
Because they're referring to a symbol that needs to be updated because the prefix itself was updated, which implies they are part of the prefix  `{ type A = AIn } --(ASF)--> { type A' = Int }` implies `{ type A = AIn }#A --(ASF)--> { type A' = Int }#A'`. Not arguing with the logging. I'll add it tomorrow.

adriaanm commented:
Closing while I build my case why the new test failures are the way things should be.

adriaanm commented:
@retronym pointed to a very relevant similar bug (https://github.com/scala/scala/pull/2986), which indicates the condition for co-evolution is still too strict

That test case has a `TypeRef(Foo.type, class Bar, Nil)` that becomes `TypeRef(AnyRef{type Bar}, type Bar, Nil)`. I'm not sure TypeMap is the right place to make this kind of decision, as in other scenarios you don't want to make that change (e.g. neg/t4137.scala, where I think a more liberal coevolveSym would erroneously rebind type aliases)

paulp commented:
Don't forget to take a gander at [SI-6161](https://issues.scala-lang.org/browse/SI-6161) armed with all these fancy new insights.

adriaanm commented:
A more ambitious refactoring: https://github.com/adriaanm/scala/compare/t1877-wip?expand=1
Not sure I can push that one through...

Not much progress yet on SI-6161 and the other related bugs (except for SI-6493, which now seems down to one problem rather than two conspirators). Will keep pulling the thread...

paulp commented:
Tsk tsk dude.

```
  object Aliases { type AThing[AIn] = Thing { type A = AIn } }
  class View[AIn](val in: Aliases.AThing[AIn]) { def f(p: in.A): in.A = p }
```

And we're back at square one. There's a ton of hard-won logic already written which accounts for this sort of thing, but it has to be used. Otherwise there will be an infinite supply of these cases where it works in the trivial example and fails when you compose two trivial examples.

I played with it a while. My way, that compiles:

```
[log typer] Thing#A is refined in both p2.Thing{type A = Int} and p2.IntThing.type, substituting A#15499 ==> A#15346
[log typer] Thing#A is refined in both p2.Thing{type A = Int} and p2.IntThing.type, substituting A#15499 ==> A#15346
[log typer] Thing#A is refined in both p2.Aliases.AThing[AIn](p2.Thing{type A = AIn}) and p2.Aliases.AThing[Int](p2.Thing{type A = Int}), substituting A#15361 ==> A#15499
[log typer] Thing#A is refined in both p2.Aliases.AThing[AIn](p2.Thing{type A = AIn}) and p2.Aliases.AThing[Int](p2.Thing{type A = Int}), substituting A#15361 ==> A#15499
[log typer] Thing#A is refined in both p2.Aliases.AThing[AIn](p2.Thing{type A = AIn}) and p2.Aliases.AThing[Int](p2.Thing{type A = Int}), substituting A#15361 ==> A#15499
```

I had a few failures. Several were neg tests which should be pos tests, like this one:

```
  // this is a compilation error at the moment as explained in SI-8104
  // because matchesPt in implicit search says that depoly(<type of materializeGeneric>) isn't a subtype of Generic.Aux[C, (Int, Int)]
  // which is rightfully so, because depoly only replaces type parameters, not type members with wildcard types
  // however in the future we might want to relax the matchesPt check, so this might start compiling
  // therefore, if you've broken this test, then you should be happy, because most likely you've just enabled an interesting use case!
```

A legitimate failure was pos/depmet_implicit_oopsla_session_simpler.scala, because it turns out there might be typevars squirreled away in the decls of the refinement type prefix which is being replaced. I saved that one by looking for typevars and declaring (pre memberType sym) =:= (newPre memberType newSym) before I threw away pre.

Another legitimate failure was one which currently fails in refchecks because the coevolved symbol is ill-kinded. An implementation artifact that the check happens at point A and not point B.

So, it's all pretty sick. I'm going to play a little more and then I'll send you my variation.

paulp commented:
Okay, it's not the prettiest thing ever but that's about 10 hours longer than I should be spending on any of this stuff so let's cut it off at that.  paulp/scala@6455c52ed6

paulp commented:
Now better known as #3501 .

adriaanm commented:
As much as this wart bothers me, I really think we should stick to the laser wart surgery at this point. This is one of the key places where we should spend some time in 2.12 to investigate the immutable approach suggested by @paulp, which is also being researched in dotty.

adriaanm commented:
More concretely, the more uniform logic is too uniform: you can't always rebind to the member found by `tp.member`, as I believe is shown by the disappearing type errors in #3501.

paulp commented:
You're probably right that it's too uniform; I still like t0764.scala to compile.

paulp commented:
"as in other scenarios you don't want to make that change (e.g. neg/t4137.scala, where I think a more liberal coevolveSym would erroneously rebind type aliases)"

Yes, exactly that happened to me initially, but with the recipe I hit upon the expanded neg/t4137.scala avoids rebinding those. Though I question whether it should have to worry about it - flush that ill-kinded shit and don't make all the other logic be paranoid and defensive. It's like letting null run amuck.

adriaanm commented:
[I think the neg/t4137 is due to using memberType (highType) rather than memberInfo (highInfo/lowInfo).] EDIT: Scratch that, it already is in this PR -- I managed to accidentally sneak those commits in here already... rebasing and then I need to look at my other bugs for a while

adriaanm committed:
SI-8177 refine embeddedSymbols

We look for any prefix that has a refinement class for a type symbol.
This includes ThisTypes, which were not considered before.

pos/t8177g.scala, neg/t0764*scala now compile, as they should

Additional test cases contributed by Jason & Paul.

adriaanm commented:
Should be good up to 600179c, but will likely have to drop 55e8551.

adriaanm commented:
integrating fix for SI-7753 from #3486 into [t8177-wip](https://github.com/adriaanm/scala/compare/scala:master...t8177-wip?expand=1), tests: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/1201/console

adriaanm commented:
review by @retronym, fixes for [SI-7753](https://github.com/adriaanm/scala/compare/scala:master...t7753-wip?expand=1), [SI-6493](https://github.com/adriaanm/scala/compare/scala:master...t6493-wip?expand=1) pending, on top of this PR.

adriaanm commented:
I've untangled my mess and will submit separate PRs for SI-7753 and SI-6493

adriaanm commented:
TODO, potentially after merging this PR (since others depend on it and we're short on time):
- [ ] turn https://github.com/paulp/scala/blob/ticket/8177/test/files/pos/t0764.scala into the corresponding neg/ test

adriaanm commented:
TODO: test/files/pos/t8177g.scala.... :(

adriaanm commented:
Closing while I investigate the new failure on the block

retronym committed:
Merge pull request #3512 from adriaanm/t261

 private vals in traits depend on composition order

adriaanm committed:
Merge pull request #3437 from Ichoran/issue/6736

SI-6736 Range.contains is wrong

adriaanm commented:
Looks like github does not like my rebasing on master.
